From 389780a70161e8591f40e61e20c232cf64268599 Mon Sep 17 00:00:00 2001
From: Jacob Young <jacobly0@users.noreply.github.com>
Date: Tue, 14 Oct 2025 02:36:07 -0400
Subject: [PATCH 025/133] Elf: fix alignment of merge subsections
X-Developer-Signature: v=1; a=openpgp-sha256; l=1116; i=xtex@astrafall.org;
 h=from:subject; bh=qEtkVkkyp5MDBoxFghlo84Ifsr6x4I7SoPQcDs43nMw=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn7cv/krU2/xxcQ2fzdc0Hl9C1e8kFnK5odJrxa92pK3
 jbDHe//dkxkYRDjYrAUU2QpMmzwZtVJ5xddVi4LM4eVCWSItEgDAwMDAwsDX25iXqmRjpGeqbah
 nqGhjpGOEQMXpwBMdVkrI8Pvk+fvOTLWXxFL5T9k27vn8GqJ1U6zjiuJeneosMaU94Yy/A9uZjT
 +dMdKiKMlOfPFU+vMHFaRqgfp5z/Yd3/Il/1bxAAA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

Closes #25565

(cherry picked from commit 6e8b07ca1526d2faa1805f2831c862b6b2337d92)
---
 src/link/Elf/Object.zig | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/link/Elf/Object.zig b/src/link/Elf/Object.zig
index a330a184bec1..1a685cd19c9c 100644
--- a/src/link/Elf/Object.zig
+++ b/src/link/Elf/Object.zig
@@ -861,7 +861,10 @@ pub fn resolveMergeSubsections(self: *Object, elf_file: *Elf) error{
         for (imsec.strings.items, imsec.subsections.items) |str, *imsec_msub| {
             const string = imsec.bytes.items[str.pos..][0..str.len];
             const res = try msec.insert(gpa, string);
-            if (!res.found_existing) {
+            if (res.found_existing) {
+                const msub = msec.mergeSubsection(res.sub.*);
+                msub.alignment = msub.alignment.maxStrict(atom_ptr.alignment);
+            } else {
                 const msub_index = try msec.addMergeSubsection(gpa);
                 const msub = msec.mergeSubsection(msub_index);
                 msub.merge_section_index = imsec.merge_section_index;
-- 
2.52.0

